### PR TITLE
feat: systemic multimodal assistant-only loss masking + cfg.role_boundaries`

### PR DESCRIPTION
**Pull request by thad0ctor**
# Description

Fixes silent ignoring of `cfg.train_on_inputs` / `cfg.roles_to_train` / `cfg.train_on_eos` in the multimodal (VLM) training path. Before this PR, only Gemma 3n honored these knobs — every other multimodal model trained on
the **entire sequence** (system + user + assistant) regardless of config, silently turning assistant-only SFT into full-sequence SFT.

### What changed

- **`ProcessingStrategy` gains a declarative boundary scanner.** Each strategy declares per-role start/end markers via `_build_role_boundaries`; a shared scanner walks the re-tokenized sequence and honors  `train_on_inputs` / `roles_to_train` / `train_on_eos` (including `"last"`).
- **New per-template strategies:** Gemma 4, Llama 3.2 Vision, Llama 4,  Pixtral, Mistral V7 Tekken.
- **Refactored:** Gemma 3 (previously no role masking), Gemma 3n (previously ad-hoc scanner → shared).
- **`cfg.role_boundaries` YAML override (opt-in)** — declare per-role markers  directly in config without subclassing. Intended as an escape hatch for  strategies whose tokens couldn't be verified offline (Voxtral, SmolVLM2,  Mistral3, InternVL, GLM4V, `llava` / `lfm2vl` / unknown template) and for  custom chat templates. Leaving the field unset (or setting it to `[]`) uses  the strategy's built-in markers — writing it literally only replaces the  built-ins when you supply a non-empty list.
- **Visibility:** fallback strategies emit a **one-shot warning** (no silent mis-masking) and every strategy logs its resolved masking config at INFO.

### Usage

Masking knobs are **per-dataset** (under `datasets:` / `test_datasets:`),not top-level. Only `train_on_inputs` lives at the root:

```yaml
train_on_inputs: false

datasets:
  - path: data/train.jsonl
    type: chat_template
    roles_to_train: [assistant]
    train_on_eos: turn         # "turn" | "all" | "none" | "last"

test_datasets:
  - path: data/val.jsonl
    type: chat_template
    split: train
    roles_to_train: [assistant]
    train_on_eos: turn
```

For a fallback strategy or custom template, add `cfg.role_boundaries` at the root (it **replaces** any built-in markers):

```yaml
# Each entry declares one role's token-level span. `role` must match the
# names you use in `roles_to_train` (e.g. assistant, user, system, tool,
# ipython). Any role not declared here is treated as masked by default.
role_boundaries:
  - role: assistant
    start: "<|turn>model"
    end: "<turn|>"
    # Optional per-entry keys (shown with defaults):
    # include_start: false        # start marker never contributes to loss
    # include_end: true           # end marker contributes on trainable turns
    #                             # (further gated by cfg.train_on_eos)

  - role: user
    start: "<|turn>user"
    end: "<turn|>"

  - role: system
    start: "<|turn>system"
    end: "<turn|>"

  - role: tool
    start: "<|turn>tool"
    end: "<turn|>"

  - role: ipython               # e.g. Llama 3.2 Vision / Llama 4 tool-call replies
    start: "<|start_header_id|>ipython<|end_header_id|>\n\n"
    end: "<|eot_id|>"

  # Special `end:` values you can mix in on any entry:
  # - role: assistant            # Pixtral-style: assistant span ends at EOS
  #   start: "[/INST]"
  #   end: eos_token             # sentinel → resolves to tokenizer.eos_token_id
  # - role: assistant            # span runs to end of sequence (no explicit close)
  #   start: "<|turn>model"
  #   end: null
```

Notes:

- `role_boundaries` is an **opt-in override**. A non-empty list **replaces**  the strategy's built-in markers 
- `start` / `end` are literal strings; they're encoded at strategy init via  `tokenizer.encode(..., add_special_tokens=False)` and the resolved token ids are logged at INFO so mismatches are visible in the training log.
- `cfg.roles_to_train` still governs which declared roles contribute to  loss. Declaring `user` / `system` / `tool` boundaries lets the scanner  correctly identify their spans as masking boundaries 
- Invalid specs (missing `role` / `start`, unencodable markers) raise at strategy init — not silently at loss-compute time.

See `docs/multimodal_assistant_mask.md` for the full audit table, per-strategy boundary markers

### Verifying it works

At startup, `build_collator` and each `ProcessingStrategy.__init__` emit  INFO lines:

```
MM collator: train_on_inputs=False roles_to_train=['assistant'] train_on_eos=turn role_boundaries_override=none
ProcessingStrategy init: class=Gemma4ProcessingStrategy ... boundaries_source=built-in boundaries=3
```

## Motivation and Context

`MultiModalChatDataCollator` re-tokenizes raw `messages` at collation time via `processor.apply_chat_template`, discarding the per-role labels computed by `ChatTemplateStrategy.tokenize_prompt` in preprocessing. It then called `processing_strategy.process_labels(input_ids)` to rebuild role-aware labels— but the base `_mask_non_assistant` was a no-op `return labels`. Only `Gemma3nProcessingStrategy` overrode it. As ar result every multimodal model except Gemma 3n, retokenized labels were **never masked by role**. 

### Offline unit tests — 63 passing

`pytest tests/test_processing_strategies.py` — no HF Hub access required.

Coverage:
- Scanner semantics: `train_on_inputs`, `roles_to_train` (incl. empty list), `train_on_eos` (`"turn"` / `"all"` / `"none"` / `"last"`), longest-prefix start match, truncated spans, scanner rewind on shared tokens.
- Per-strategy masking with fake tokenizers mirroring real-model id layouts:  Qwen2-VL, Qwen3.5, Gemma 3 / 3n / 4, Llama 3.2 Vision, Llama 4, Pixtral,  Mistral V7 Tekken.
- Media-token masking inside assistant spans (`image_pad`, `video_pad`,  `boi`/`eoi`/`boa`/`eoa` where applicable).
- Dispatcher routing including Mistral3 / GLM4V lazy-import fallbacks.
- `cfg.role_boundaries` override: non-empty replaces built-ins wholesale,  empty list falls through to built-ins (opt-in semantics), enables  unverified strategies, `eos_token` sentinel, `null` end, spec-validation  errors, Pydantic model input.
- `SFTDataset` schema accepts every value the scanner honors for `train_on_eos` — `"all"` / `"turn"` / `"last"` / `"none"` — and rejects  bogus values (regression against a prior `Literal` that dropped `"none"`).
- Edge cases: batch_size > 1, all-pad sequences, consecutive assistant turns, pad masking inside trainable spans.
- Init-time INFO log is grep-visible (regression test with `caplog`).

### End-to-end against real tokenizers

- **`google/gemma-4-E2B-it`**: 13/40 tokens kept for a 2-turn chat; decoded preview shows only assistant responses + `<turn|>` markers retained.
- **`axolotl-ai-co/Llama-3.3-70B-Instruct-tokenizer`** with bundled  `llama3_2_vision.jinja`: 11/64 tokens kept; decoded content resolves to  `"The capital of France is Paris.<|eot_id|>"` and `"Berlin.<|eot_id|>"`.- Gemma 4 boundary ids verified against the real tokenizer:  `<|turn>model` → `[105, 4368]`, `<turn|>` → `[106]`, `<|image|>` →  `258880`, `<|audio|>` → `258881`, `<|video|>` → `258884`.

### Real training run

Live SFT on **Gemma 4 E2B**, single RTX 5090, FSDP2 + CPU offload,
`lora_r=256` rsLoRA, `sequence_len=5048`:

- Collator log confirms knobs threaded through:
  ```
  MM collator: train_on_inputs=False roles_to_train=['assistant'] train_on_eos=turn role_boundaries_override=none
  ProcessingStrategy init: class=Gemma4ProcessingStrategy ... boundaries_source=built-in boundaries=3
  ```
- Per-step counters confirm masking is active:  `tokens/total: 59664, tokens/trainable: 13945` (~23.4% unmasked —
  consistent with assistant-only masking on an OCR conversation where user content dominates).
- Training + eval progress cleanly; no "legacy behavior" warnings.

### Back-compat

- **Text-only path unaffected** — MM path only taken when  `cfg.processor_type and self.processor`.
- **No schema break** — defaults when the per-dataset keys are unset match  the text-only `ChatTemplateStrategy` defaults (`roles_to_train=["assistant"]`,  `train_on_eos="turn"`).
- **Loss values will change** for any multimodal run that was previously  seeing full-sequence loss — this is the intended fix. Sample-efficiency  should improve; compare eval loss/perplexity on a held-out set to confirm.

## AI Usage Disclaimer

Yes. Claude Opus 4.7 (via Claude Code) assisted throughout. All commits are co-authored and were reviewed against real training logs by the submitter before inclusion and changes and testing reviewed and tested by Codex as well.Design choices and approach human guided.

## Types of changes

- [x] **Bug fix** (non-breaking change which fixes an issue) — restores honoring of `train_on_inputs` / `roles_to_train` / `train_on_eos` in     the multimodal path.
- [x] **New feature** (non-breaking change which adds functionality) —  `cfg.role_boundaries` YAML override, new per-template strategies, and INFO-log visibility of resolved masking config.
- [x] **Documentation update** — new `docs/multimodal_assistant_mask.md`
      (audit table, root cause, design, config guide).

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed multimodal training where token masking settings could be ignored during collation by ensuring dataset-level masking params are respected.

* **New Features**
  * Role-boundary token masking for precise control of which tokens contribute to loss.
  * Optional config override to supply custom role boundary markers.
  * Added "none" option to disable training on EOS tokens.

* **Documentation**
  * New guidance on configuring, verifying, and logging multimodal masking behavior.

* **Tests**
  * Comprehensive tests covering masking logic, EOS modes, overrides, and strategy routing.

### CODE REVIEW

**Comment by thad0ctor** [comment]
Apologies in advance that this is a significant Opus-written PR, I can architect but my coding lacks.  All +30 local fork PRs were reviewed by coderabbit and codex as well and tested thoroughly with training runs and against the existing workflow tests in the repo in my fork.

File: src/axolotl/processing_strategies.py [hunk 1]
```diff
@@ -221,57 +294,287 @@ def convert_messages_to_multimedia_messages(messages: list[dict]) -> list[dict]:
...
+                    and train_on_eos not in ("none", "last")
+                ):
+                    mask[i][content_end:end_after] = 1
+                if found_end and best_match.include_end and train_on_eos == "last":
+                    last_trainable_end_span[i] = (content_end, end_after)
+            else:
+                # Non-trainable role: only the end marker can contribute, and only on train_on_eos="all".
+                if found_end and train_on_eos == "all":
+                    content_end = end_after - len(best_match.end_tokens)
+                    mask[i][content_end:end_after] = 1
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply src/axolotl/processing_strategies.py:418]: Fixed and added regression test for this case

File: src/axolotl/processing_strategies.py [hunk 2]
```diff
@@ -280,141 +583,309 @@ def __init__(
...
+            role_boundaries_override=role_boundaries_override,
         )
+        # Gemma3 uses boi_token as the image placeholder.
+        special_tokens_map = (
+            getattr(processor.tokenizer, "special_tokens_map", {}) or {}
+        )
+        boi = special_tokens_map.get("boi_token")
+        if boi is not None:
+            self.image_token = boi
+            self.image_token_id = processor.tokenizer.convert_tokens_to_ids(boi)
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply src/axolotl/processing_strategies.py]: Valid and addressed in subsequent commit

File: src/axolotl/processing_strategies.py [hunk 3]
```diff
@@ -499,29 +1029,48 @@ def __init__(
...
         return labels
 
 
 class Glm4vProcessingStrategy(ProcessingStrategy):
-    """Processing Strategy class for GLM4V and GLM4V-MoE vision models."""
+    """Processing Strategy class for GLM4V / GLM4V-MoE.
+
+    Role boundaries NOT declared — GLM4V markers (``<|assistant|>`` /
+    ``<|user|>``) unverified against a real checkpoint.
+    """
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply src/axolotl/processing_strategies.py:1055]: Noted, reviewed against models , will address with in follow on commit with test coverage

File: docs/multimodal_assistant_mask.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    train_on_eos: turn
+```
+
+### How to verify at runtime
+
+`build_collator` logs the resolved knobs at INFO:
+
+```
+MM collator: train_on_inputs=False roles_to_train=['assistant'] train_on_eos=turn role_boundaries_override=none
+```
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply docs/multimodal_assistant_mask.md:31]: Will cleanup as part of pending commit to shorten docstrings /descriptions

**Review by winglian — APPROVED** [review]

**Comment by thad0ctor** [comment]
@winglian thank you!

**Comment by NanoCode012** [comment]
Thanks for the PR! I think the next step from here would be to review speed / performance cost for this change and whether the implementation can be improved upon.

**Comment by thad0ctor** [comment]
> Thanks for the PR! I think the next step from here would be to review speed / performance cost for this change and whether the implementation can be improved upon.

will do, I'll look into benchmarking on my end once my GPUs free up

**Comment by thad0ctor** [comment]
> Thanks for the PR! I think the next step from here would be to review speed / performance cost for this change and whether the implementation can be improved upon.

I benched the PR, the scanner cost is ~0.21 µs/token (linear, strategy-independent), which works out to ~7 ms at micro batch=8 × seq=4096, well under 0.1% of step time and fully overlapped with GPU compute via dataloader workers. 

  Digging deeper, there are two possible directions that seem viable if you'd prefer without tapping into preprocessing (although that could be a viable future decision) to reduce any latency all together: 
                                                                                                                                                                                        
  (a) **Vectorized scanner:** I tested this on a local branch (_apply_role_boundaries_vectorized alongside the original): ~1.5–2× faster (Gemma4 1.64×, Llama3.2V 1.50×, Pixtral 1.93×), this approach tested well and seems valid. 
                                                                                                                                                                                        
  (b) **Native return_assistant_tokens_mask=True from apply_chat_template:** HF's tokenizers/processors expose this flag for templates with {% generation %} blocks.  Where supported, the tokenizer returns the assistant mask directly, eliminating the scanner, the per-strategy boundary declarations, and the cfg.role_boundaries override . This is closer in spirit to the text path (prompt_strategies/chat_template.py), which defers boundary detection to template machinery (find_turn does a template-diff at preprocess time) rather than running its own per-batch token scan. The would would probably be: detect support at strategy init, use native when available, current scanner as fallback. A bit more effort, but architecturally consistent.    
  
 My take is that the scanner affords maximum flexibility at a noise-level latency tax. Improving it with the vectorized approach makes it a bit more palatable; frankly multimodal training is so much lengthier than text that a <1% tax is an acceptable trade-off pending a more complex refactor. To make (b) work with the same flexibility (given the diversity of chat templates for models without {% generation %} blocks) would be a strong contender for a 3rd future commit: native mask as a fast path for the supported common case (roles_to_train=["assistant"], train_on_eos="turn", template has generation blocks), scanner as the fallback for multi-role training, non-default EOS modes, cfg.role_boundariesoverrides, and templates without generation blocks. 
  
  That said, in hindsight perhaps (b) would have been more appropriate; I know your strategy has been to enforce a standard chat template (and rightfully so if you look at the complexity of chat template handling in llama.cpp and the like) and this deviates from it, so part of this may have been my own preferences leaking in.

**Comment by NanoCode012** [comment]
You can try timing for option b, but like mentioned earlier, it was very slow when I first tested it.
